File: mod_data/miney/channel.lua
-- The file channel: how Miney's Python side reaches this mod without being a player.
--
-- Two append-only logs in this mod's own data directory, one JSON record per line.
-- Python appends a request to `c2s`, a globalstep here reads whatever complete lines
-- have appeared since the last one, and answers go out as lines on `s2c`. That is the
-- whole transport: no account, no port, no password, and it works in a singleplayer
-- world started from the Luanti menu, where a second network client is refused outright.
--
-- Three things about it are not obvious and were measured rather than assumed:
--
--   * `io.open` runs the engine's security path check, and on a *read* that resolves
--     builtin, the game directory and every loaded mod path. 2374 microseconds per call
--     on Windows with 34 mods loaded; 214 on Linux. `seek` on a handle that is already
--     open costs 1.9. So both handles are opened once, while mods load, and never
--     again - an idle channel costs one `seek("end")` per step.
--   * `flush()` is not a disk write. It moves bytes out of stdio into the kernel, which
--     is what makes them visible to the other process; the page cache answers Python's
--     read. `core.safe_file_write` would be the obvious call here and is the wrong one:
--     it fsyncs, 4537 microseconds every time.
--   * The newline is the commit marker. A record without its terminator is not a short
--     record, it is not a record at all - the reader leaves its offset where it is and
--     sees the whole line next step. That is what makes a half-written request
--     harmless instead of merely unlikely.
--
-- And the rule that cost a frozen server to find: never re-check the input inside the
-- drain loop. One snapshot per step, process what is in it, return. A loop that looks
-- again keeps finding more, because a client refills faster than a path-checked open
-- drains, and the globalstep never returns.

local modname = minetest.get_current_modname()

miney_channel = {}

-- How long a session may go without saying anything before the mod throws away
-- everything it left behind. Python sends a ping well inside this; the timeout is for
-- the process that was killed rather than the one that said goodbye.
local SESSION_TIMEOUT = 30

-- How much of one server step the pump may spend running requests. A thousand trivial
-- commands cost 3 ms and a thousand node placements do not, so the batch is bounded by
-- the clock and the rest waits for the next step. Lowering it makes a busy Miney script
-- gentler on everyone else in the world.
local STEP_BUDGET_US = tonumber(minetest.settings:get("miney_channel_budget_ms") or "5") * 1000

-- The channel directory, one per world.
--
-- `get_mod_data_path()` is per mod but not per world, and two servers may well be up at
-- once - a project's world and something the user started from the menu. Sharing one
-- `c2s` between them would interleave two request streams into one file. So each world
-- gets its own subdirectory, named after it and stamped with a digest of its full path
-- because two worlds may share a name.
--
-- Both functions only answer while mods load, which is why this runs at dofile time.
local function world_key()
    local path = minetest.get_worldpath()
    local name = string.match(path, "([^/\\]+)[/\\]*$") or "world"
    name = string.gsub(name, "[^%w_%-]", "_")
    local digest
    if minetest.sha1 then
        digest = string.sub(minetest.sha1(path), 1, 8)
    else
        local hash = 5381
        for index = 1, #path do
            hash = (hash * 33 + string.byte(path, index)) % 4294967296
        end
        digest = string.format("%08x", hash)
    end
    return name .. "-" .. digest
end

local base = (minetest.get_mod_data_path and minetest.get_mod_data_path())
    or (minetest.get_worldpath() .. DIR_DELIM .. "miney_channel")
minetest.mkdir(base)

local dir = base .. DIR_DELIM .. world_key()
minetest.mkdir(dir)

local c2s_path = dir .. DIR_DELIM .. "c2s"
local s2c_path = dir .. DIR_DELIM .. "s2c"
local beacon_path = dir .. DIR_DELIM .. "beacon.json"

local requests = nil    -- read handle on c2s, held for the life of the server
local answers = nil     -- append handle on s2c, likewise
local read_off = 0      -- how far into c2s this mod has got
local dirty = false     -- something was written to s2c and not yet flushed

local sessions = {}     -- id -> last time anything arrived from it
local dispatch = nil    -- what init.lua does with a fields table
local forget = nil      -- what init.lua does when a session goes away
local advertised_api = 0 -- the contract version, sent in the beacon and every hello

local function log(level, message)
    minetest.log(level, "[" .. modname .. "/channel] " .. message)
end

--- Send one answer to a session. The only way anything leaves this mod.
--
-- Written straight into the stdio buffer and flushed once at the end of the step, so a
-- batch of a thousand commands costs one flush rather than a thousand.
--
-- Global rather than a field of miney_channel, because every other file in the mod -
-- the Lua result, the callback acknowledgement, the event - reaches for it by name.
--
-- @param who (string) - The session name: "@" and its id, as handle_fields was given
--     it. The "@" keeps a session from ever colliding with a player name, which Luanti
--     builds out of letters, digits, dash and underscore only.
-- @param tbl (table) - The payload.
function miney_reply(who, tbl)
    if not answers then
        return
    end
    local ok, line = pcall(minetest.write_json, {session = string.sub(who, 2), data = tbl})
    if not ok or line == nil then
        line = minetest.write_json({
            session = string.sub(who, 2),
            data = {error = "Miney could not turn this answer into JSON: " .. tostring(line)},
        })
    end
    -- write_json escapes newlines inside strings, so this never fires. It is here
    -- because a raw newline would not shorten a record, it would split it into two
    -- unparseable ones and desynchronise the log for the rest of the session.
    line = string.gsub(line, "[\r\n]", " ")
    answers:write(line, "\n")
    dirty = true
end

local function drop_session(id, why)
    sessions[id] = nil
    if forget then
        forget("@" .. id)
    end
    log("action", "Session " .. id .. " " .. why .. ".")
end

-- One record from c2s. Everything it can be wrong about is answered on s2c, because a
-- request that vanishes without a word is the one failure a beginner cannot debug.
local function handle_record(line)
    -- An empty line is what a client sends when it attaches, to close off a record
    -- some earlier process died halfway through. Nothing arrived and nothing is
    -- wrong, so it must not reach parse_json - which would print three lines of
    -- engine error into the server log for every ordinary connect.
    if line == "" or string.match(line, "^%s*$") then
        return
    end

    local ok, record = pcall(minetest.parse_json, line)
    if not ok or type(record) ~= "table" then
        log("warning", "Ignored an unparseable request.")
        return
    end
    local id = record.session
    if type(id) ~= "string" or id == "" then
        log("warning", "Ignored a request without a session.")
        return
    end

    local known = sessions[id] ~= nil
    sessions[id] = os.time()

    if record.op == "bye" then
        drop_session(id, "said goodbye")
        return
    end
    if record.op == "hello" then
        if known then
            -- A second hello under the same id is a Python process that restarted
            -- without saying goodbye. Its old sandbox and its old timers are not
            -- wanted; the id is.
            if forget then forget("@" .. id) end
        end
        miney_reply("@" .. id, {ok = true, action = "hello", mod_api = advertised_api})
        return
    end
    if record.op == "ping" then
        return
    end
    if type(record.fields) == "table" and dispatch then
        dispatch("@" .. id, record.fields)
    end
end

-- Everything that has appeared in c2s since the last step, up to the time budget.
--
-- Deliberately one read of one snapshot. read_off only ever moves past a line that had
-- its terminator, so a request caught mid-write is simply seen whole next time.
local function pump()
    if not requests then
        return
    end

    local size = requests:seek("end")
    if size < read_off then
        -- The request log got shorter than this mod has already read, which means
        -- something started it over behind our back. Reading on from the old offset
        -- would land in the middle of a record, so start over too.
        read_off = 0
        log("action", "The request log was started over; reading it from the beginning.")
    end
    if size == read_off then
        return
    end

    requests:seek("set", read_off)
    local chunk = requests:read("*a") or ""

    local deadline = minetest.get_us_time() + STEP_BUDGET_US
    local pos = 1
    while true do
        local stop = string.find(chunk, "\n", pos, true)
        if not stop then
            break
        end
        handle_record(string.sub(chunk, pos, stop - 1))
        pos = stop + 1
        if minetest.get_us_time() > deadline then
            break
        end
    end
    read_off = read_off + (pos - 1)
end

local function expire_sessions()
    local now = os.time()
    for id, seen in pairs(sessions) do
        if now - seen > SESSION_TIMEOUT then
            drop_session(id, "went away without saying goodbye")
        end
    end
end

--- Start the channel.
--
-- @param dispatch_fn (function) - Called as (who, fields) with a request's fields.
-- @param forget_fn (function) - Called as (who) when a session ends, for cleanup.
-- @param mod_api (number) - The contract version to advertise in the beacon.
function miney_channel.start(dispatch_fn, forget_fn, mod_api)
    dispatch = dispatch_fn
    forget = forget_fn
    advertised_api = mod_api

    minetest.register_on_mods_loaded(function()
        -- Both logs are emptied here and nowhere else, and this is the one moment when
        -- that is unambiguous: nothing is attached to a server that has not finished
        -- loading, so neither file has a reader whose offset could be left pointing
        -- into the middle of a record.
        --
        -- ponytail: that also puts a ceiling on how large they get - one server uptime,
        -- not one machine's lifetime. Compacting them while a session is running needs
        -- a handshake in both directions, because either side truncating a file the
        -- other is reading is a race that a poll interval only usually wins. If a
        -- long-running server ever fills a disk with these, that handshake is the fix;
        -- until then the restart is.
        for _, path in ipairs({c2s_path, s2c_path}) do
            local blank = io.open(path, "w")
            if blank then blank:close() end
        end
        requests = io.open(c2s_path, "r")
        answers = io.open(s2c_path, "a")
        if not requests or not answers then
            log("error", "Could not open the channel files in " .. dir ..
                ". Miney's Python side will not be able to reach this server.")
            return
        end
        read_off = 0

        -- The beacon is how Python finds this server at all: it knows path_user and
        -- nothing else. One fsync per server start is a price worth paying for a file
        -- that must survive a crash intact.
        minetest.safe_file_write(beacon_path, minetest.write_json({
            mod_api = mod_api,
            engine = (minetest.get_version() or {}).string or "",
            world = minetest.get_worldpath(),
            world_name = string.match(minetest.get_worldpath(), "([^/\\]+)[/\\]*$") or "",
            dir = dir,
            singleplayer = minetest.is_singleplayer(),
            started = os.time(),
        }))
        log("action", "Channel open at " .. dir .. ".")
    end)

    minetest.register_globalstep(function()
        pump()
        if dirty then
            answers:flush()
            dirty = false
        end
    end)

    -- Once a second is often enough for a timeout measured in tens of them, and it
    -- keeps the per-step cost at the one seek the pump already does.
    local function tick()
        expire_sessions()
        minetest.after(1, tick)
    end
    minetest.after(1, tick)

    minetest.register_on_shutdown(function()
        -- A beacon left behind by a killed server is exactly what Python cannot tell
        -- from a live one, so the orderly case at least cleans up after itself.
        os.remove(beacon_path)
        if answers then answers:flush() end
    end)
end

return miney_channel

File: miney/channel.py
"""
How Miney reaches the Luanti server.

The Lua mod keeps an append-only log in Luanti's own ``mod_data`` directory, one for
each direction, and moves everything across on the server step. This module is the
other end of it. There is no account, no port, no password and nothing to configure:
if a server with the Miney mod is running on this computer, the beacon it wrote says
where its channel is, and appending a line to that channel runs Lua on it.

That also covers the case no network client can reach. A singleplayer world started
from the Luanti menu refuses a second client outright
(``serverpackethandler.cpp:152-158``), and a file is not a client.

Nothing in here is part of the public API. :class:`~miney.luanti.Luanti` opens the
channel, and the words "channel", "session" and "offset" never reach a script.
"""
from __future__ import annotations

import json
import logging
import os
import sys
import threading
import time
import uuid
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable

from .exceptions import MineyRunError
# The module, not the function: the test suite redirects `default_luanti_dir` away from
# a real home directory by patching it there, and a name bound here at import time
# would not see that.
from .env import paths as env_paths

logger = logging.getLogger(__name__)

#: Environment variables that move Luanti's user directory, in the order the engine
#: reads them (``porting.cpp:172-180``). The second one is deprecated and still honoured.
USER_PATH_VARS = ("LUANTI_USER_PATH", "MINETEST_USER_PATH")

#: How often the reader thread looks for new answers. The floor for a round-trip is the
#: server step - 17 ms in a game you are standing in, 31 on a world ``miney start``
#: launched - so polling faster than this buys nothing and this fast costs a ``stat``.
POLL_INTERVAL = 0.005

#: How often a session says it is still there. The mod forgets a session that has been
#: quiet for 30 seconds, and forgetting means cancelling its timers and dropping its
#: chat commands, so this has to be comfortably inside that.
PING_INTERVAL = 5.0


# One world has one request log, so two Miney scripts on the same world append to the
# same file. Linux makes that safe on its own: an `O_APPEND` write to a regular file
# picks its offset and writes under the same lock, so it cannot be split or landed on.
# Windows only pretends to - its C runtime implements append as a seek to the end
# followed by a write, and nothing holds those two together. Two processes can therefore
# choose the same offset and the second one writes over the first.
#
# What that costs was measured with the test below, 300 requests of 20 KB from two
# processes: on Linux 300 arrive with this lock removed, three runs out of three. On
# Windows 282 arrive and 18 are simply gone - not damaged, not short, absent, along with
# every byte of them. The script that sent them waits out its timeout for an answer to a
# request the server never saw, and the seam where the overwrite lands leaves one line
# the mod cannot parse either.
#
# Hence a real lock, taken around every write. It lives on a file of its own and never
# on `c2s`: Windows file locks are mandatory rather than advisory, so a range locked on
# the request log would make the mod's own read of it fail - the opposite of the point.
if sys.platform == "win32":
    import msvcrt

    def _lock(handle) -> None:
        """
        Take the channel's write lock, waiting for whoever has it.

        :param handle: The open lock file.
        """
        handle.seek(0)
        # LK_LOCK retries for ten seconds and then raises, which is the right shape:
        # the lock is held for the length of one write, so anything near that long is
        # a stuck process rather than contention, and the caller turns it into a
        # logged failure instead of writing into the middle of somebody else's line.
        msvcrt.locking(handle.fileno(), msvcrt.LK_LOCK, 1)

    def _unlock(handle) -> None:
        """
        Give the write lock back.

        :param handle: The open lock file.
        """
        handle.seek(0)
        msvcrt.locking(handle.fileno(), msvcrt.LK_UNLCK, 1)

else:
    import fcntl

    def _lock(handle) -> None:
        """
        Take the channel's write lock, waiting for whoever has it.

        :param handle: The open lock file.
        """
        fcntl.flock(handle.fileno(), fcntl.LOCK_EX)

    def _unlock(handle) -> None:
        """
        Give the write lock back.

        :param handle: The open lock file.
        """
        fcntl.flock(handle.fileno(), fcntl.LOCK_UN)


def _platform_user_path() -> Path | None:
    """
    Where a system-wide Luanti keeps its user directory on this platform.

    Read out of ``porting.cpp``: ``%APPDATA%\\Minetest`` on Windows,
    ``~/.minetest`` on Linux and ``~/Library/Application Support/minetest`` on macOS.
    The directory is still called *minetest* on all three - the rename has not reached
    it yet - which is exactly the kind of thing worth not guessing.

    :return: The directory, or None if the platform is unknown.
    """
    if sys.platform == "win32":
        appdata = os.environ.get("APPDATA")
        return Path(appdata) / "Minetest" if appdata else None
    if sys.platform == "darwin":
        return Path.home() / "Library" / "Application Support" / "minetest"
    return Path.home() / ".minetest"


def candidate_user_paths(luanti_dir: Path | None = None) -> list[Path]:
    """
    Every directory a running Luanti might be using as its user path.

    In the engine's own order: the environment variables first, then the Luanti that
    Miney downloads - which is portable, so its user path is the install directory
    itself - and finally the place a Luanti installed by the system would use.

    :param luanti_dir: The Miney-managed install, if one is known. Defaults to
        :func:`~miney.env.paths.default_luanti_dir`.
    :return: The candidates, most specific first, without duplicates.
    """
    found: list[Path] = []
    for variable in USER_PATH_VARS:
        value = os.environ.get(variable)
        if value:
            found.append(Path(value))
    found.append(luanti_dir or env_paths.default_luanti_dir())
    system = _platform_user_path()
    if system:
        found.append(system)

    unique: list[Path] = []
    for path in found:
        if path not in unique:
            unique.append(path)
    return unique


@dataclass(frozen=True)
class Beacon:
    """
    What a running server says about itself in ``beacon.json``.

    Written once when the mod loads and removed when the server shuts down properly.
    A server that was killed leaves it behind, and there is no process id in it to
    check against - the Lua sandbox has none to offer - so a beacon means "a server
    was here", not "a server is here". Liveness is probed, never read.

    :param directory: The channel directory this beacon was found in.
    :param mod_api: The mod's contract version, compared against
        :data:`~miney.lua.REQUIRED_MOD_API`.
    :param engine: The Luanti version string, e.g. ``"5.16.1"``.
    :param world: Full path of the world the server is running.
    :param world_name: Just the world's directory name.
    :param singleplayer: True when the server is a singleplayer game rather than a
        dedicated one.
    :param started: Unix time the mod loaded.
    """

    directory: Path
    mod_api: int
    engine: str
    world: str
    world_name: str
    singleplayer: bool
    started: int

    def __repr__(self) -> str:
        return f'<Miney channel for world "{self.world_name}" ({self.engine})>'


def read_beacon(path: Path) -> Beacon | None:
    """
    Read one ``beacon.json``.

    :param path: The file to read.
    :return: The beacon, or None if it is missing or not readable as one.
    """
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None
    if not isinstance(data, dict) or "mod_api" not in data:
        return None
    return Beacon(
        directory=path.parent,
        mod_api=int(data.get("mod_api", 0)),
        engine=str(data.get("engine", "")),
        world=str(data.get("world", "")),
        world_name=str(data.get("world_name", "")),
        singleplayer=bool(data.get("singleplayer", False)),
        started=int(data.get("started", 0)),
    )


def find_beacons(luanti_dir: Path | None = None) -> list[Beacon]:
    """
    Every Miney channel on this computer, newest first.

    One world can have one channel, and a computer can have several worlds up at once -
    the world a project started and one the user opened from the Luanti menu - so this
    can return more than one. :class:`~miney.luanti.Luanti` says so rather than picking.

    :param luanti_dir: The Miney-managed install, if one is known.
    :return: The beacons found, the most recently started one first.
    """
    beacons: list[Beacon] = []
    seen: set[Path] = set()
    for user_path in candidate_user_paths(luanti_dir):
        root = user_path / "mod_data" / "miney"
        if not root.is_dir():
            continue
        for entry in sorted(root.iterdir()):
            beacon_file = entry / "beacon.json"
            if entry.resolve() in seen or not beacon_file.is_file():
                continue
            seen.add(entry.resolve())
            beacon = read_beacon(beacon_file)
            if beacon is not None:
                beacons.append(beacon)
    beacons.sort(key=lambda one: one.started, reverse=True)
    return beacons


class FileChannel:
    """
    One session on one world's channel.

    Reached through :class:`~miney.luanti.Luanti`, never built by hand. It offers two
    things - send a field table, hear about the dictionaries that come back - and hides
    the fact that both are lines in a file.

    There is no size limit to know about: 8 MB crossed the channel in a single server
    step, indistinguishable from a kilobyte.

    :param directory: The channel directory, from a :class:`Beacon`.
    :param timeout: How long :meth:`attach` waits for the server to answer.
    """

    def __init__(self, directory: Path, timeout: float = 10.0):
        self.directory = Path(directory)
        #: Which contract version the mod on the other end speaks. Known before the
        #: first command here, because the answer to the session's hello carries it.
        self.mod_api: int | None = None
        self._session = str(uuid.uuid4())
        self._listeners: list[Callable[[dict], None]] = []
        self._send_lock = threading.Lock()
        self._running = True
        self._reader: threading.Thread | None = None
        self._last_ping = 0.0

        c2s = self.directory / "c2s"
        s2c = self.directory / "s2c"
        try:
            # Created rather than assumed: the mod makes both files while it loads, but
            # a beacon can outlive a wiped directory and a missing file must not become
            # a traceback three frames deep in a thread.
            c2s.touch(exist_ok=True)
            s2c.touch(exist_ok=True)
            self._out = open(c2s, "ab", buffering=0)
            self._in = open(s2c, "rb")
            # Never read or written, only locked. It exists so that the lock has
            # somewhere to live that the mod does not touch.
            self._lock_file = open(self.directory / "lock", "ab", buffering=0)
        except OSError as error:
            raise MineyRunError(
                f"Miney found a Luanti server at {self.directory} but cannot write to "
                f"its channel: {error}"
            ) from error

        # Start at the end of what is already there. Everything before this belongs to
        # a session that has ended, and replaying it would answer requests nobody made.
        self._offset = s2c.stat().st_size

        self.attach(timeout)

    # -- attaching ---------------------------------------------------------------

    def attach(self, timeout: float) -> None:
        """
        Announce the session and wait for the server to answer it.

        The lone newline is not decoration. A Python process killed halfway through a
        write leaves a line without its terminator, and the mod is waiting for that
        terminator before it reads anything further. Sending one turns the remains into
        a single unparseable record, which the mod logs and drops, instead of letting
        it be spliced onto the front of the first real request.

        :param timeout: Seconds to wait for the answer.
        :raises MineyRunError: If nothing answers in time.
        """
        acknowledged = threading.Event()

        def note(record: dict) -> None:
            if record.get("action") == "hello":
                acknowledged.set()

        self._listeners.append(note)
        self._reader = threading.Thread(
            target=self._read_loop, name="miney-channel-reader", daemon=True
        )
        self._reader.start()

        self._raw_write(b"\n")
        self._raw_send({"session": self._session, "op": "hello"})

        if not acknowledged.wait(timeout):
            self.close()
            raise MineyRunError(
                f"A Luanti server left its mark in {self.directory}, but it did not "
                f"answer within {timeout:.0f} seconds.\n"
                f"{self.timeout_hint()}"
            )
        self._listeners.remove(note)

    def timeout_hint(self) -> str:
        """
        What to tell somebody whose command went unanswered.

        Two things look exactly alike from here, and the likely one comes first. A
        singleplayer game with the menu open is not running at all: the server returns
        immediately at ``dtime == 0`` (``server.cpp:684-685``), so no mod code runs,
        no timer fires and nothing arrives until the menu closes. It is invisible from
        the outside, and pressing ESC to go back to the editor is the most ordinary
        thing a beginner does.

        :return: A sentence naming both possibilities and what to do about them.
        """
        return (
            "If the game is open in a window, check whether it is paused - a "
            "singleplayer world stops completely while the ESC menu is up, and nothing "
            "Miney sends is looked at until you close it. Otherwise the server is no "
            "longer running and has to be started again."
        )

    # -- the transport interface -------------------------------------------------

    @property
    def connected(self) -> bool:
        """Whether this session is still usable."""
        return self._running

    def add_listener(self, listener: Callable[[dict], None]) -> None:
        """
        Hear about every answer and every event the server sends.

        :param listener: Called with each decoded record, on the reader thread.
        """
        self._listeners.append(listener)

    def send(self, fields: dict[str, str]) -> bool:
        """
        Send one request.

        :param fields: The request.
        :return: True if it was written.
        """
        return self._raw_send({"session": self._session, "fields": fields})

    def close(self) -> None:
        """
        End the session and let the server clean up after it straight away.

        Without the goodbye the mod waits out its own timeout before cancelling
        whatever timers and chat commands this session left behind, which is half a
        minute of a script that has ended still doing things.
        """
        if not self._running:
            return
        try:
            self._raw_send({"session": self._session, "op": "bye"})
        except Exception:  # noqa: BLE001 - shutting down, nothing left to tell
            pass
        self._running = False
        if self._reader is not None and self._reader.is_alive():
            self._reader.join(timeout=1.0)
        for handle in (self._out, self._in, self._lock_file):
            try:
                handle.close()
            except OSError:
                pass

    # -- the machinery ------------------------------------------------------------

    def _raw_send(self, record: dict[str, Any]) -> bool:
        """
        Write one record as a line.

        :param record: What to send.
        :return: True if it was written.
        """
        return self._raw_write(
            json.dumps(record, ensure_ascii=False).encode("utf-8") + b"\n"
        )

    def _raw_write(self, blob: bytes) -> bool:
        """
        Put bytes on the request log, whoever else is writing to it.

        Two locks, and both are needed. The threading one keeps this process's own
        threads apart - the reader sends pings from its own thread while the main one
        sends commands. The file lock keeps *other* Miney processes apart, which is a
        different problem with a different answer; see the comment on :func:`_lock`.

        :param blob: The bytes to append, terminator included.
        :return: True if they were written.
        """
        try:
            with self._send_lock:
                _lock(self._lock_file)
                try:
                    self._out.write(blob)
                finally:
                    _unlock(self._lock_file)
        except OSError as error:
            logger.error("Could not write to the Miney channel: %s", error)
            return False
        return True

    def _read_loop(self) -> None:
        """Watch the answer log and hand every complete line to the listeners."""
        partial = b""
        while self._running:
            try:
                size = os.fstat(self._in.fileno()).st_size
                if size < self._offset:
                    # The answer log got shorter than what has already been read, so a
                    # server restarted and emptied it. Reading on from the old offset
                    # would land inside a record.
                    self._offset = 0
                    partial = b""
                if size > self._offset:
                    self._in.seek(self._offset)
                    chunk = self._in.read(size - self._offset)
                    self._offset += len(chunk)
                    partial += chunk
                    # A trailing fragment is not a short record, it is not a record
                    # yet. Keeping it costs one buffer and makes a torn read impossible
                    # rather than merely unlikely.
                    *lines, partial = partial.split(b"\n")
                    for line in lines:
                        self._deliver(line)
            except (OSError, ValueError) as error:
                # ValueError is close() winning the race against this thread's next
                # read, which is an ordinary shutdown and not worth a line in the log.
                if self._running:
                    logger.error("The Miney channel stopped being readable: %s", error)
                self._running = False
                return

            now = time.monotonic()
            if now - self._last_ping > PING_INTERVAL:
                self._last_ping = now
                self._raw_send({"session": self._session, "op": "ping"})
            time.sleep(POLL_INTERVAL)

    def _deliver(self, line: bytes) -> None:
        """
        Decode one line and give it to whoever is listening.

        :param line: One record, without its terminator.
        """
        if not line.strip():
            return
        try:
            record = json.loads(line.decode("utf-8"))
        except (ValueError, UnicodeDecodeError):
            logger.warning("Ignored an unreadable answer from the server.")
            return
        if not isinstance(record, dict) or record.get("session") != self._session:
            return
        data = record.get("data")
        if not isinstance(data, dict):
            return
        if "mod_api" in data:
            self.mod_api = data["mod_api"]
        for listener in list(self._listeners):
            try:
                listener(data)
            except Exception as error:  # noqa: BLE001 - a broken listener is not fatal
                logger.error("Error handling a server answer: %s", error, exc_info=True)

    def __repr__(self) -> str:
        return f'<Miney file channel "{self.directory.name}">'
